src/Api.php

```diff
@@ -145,25 +166,179 @@ protected function computeHmac($hmac, $fields)
     /**
      * Makes an array of parameters become a querystring like string.
      *
-     * @param  array $array
-     *
      * @return string
      */
-    static public function stringify(array $array)
+    public static function stringify(array $array)
     {
-        $result = array();
+        $result = [];
         foreach ($array as $key => $value) {
             $result[] = sprintf('%s=%s', $key, $value);
         }
+
         return implode('&', $result);
     }
 
+    public function getOptions(): array
+    {
+        return $this->options;
+    }
+
     /**
-     * @return array
+     * @See documentation ManuelIntegrationVerifone_PayboxSystem_V8 5.3.4 Values checking
+     *
+     * @param GetHttpRequest $httpRequest
+     *
+     * @return bool
      */
-    public function getOptions()
+    public function verify($httpRequest)
     {
-        return $this->options;
+        $context = [
+            'httpRequestUri' => $httpRequest->uri,
+            'httpRequestQuery' => $httpRequest->query,
+        ];
+        if (false === $this->doesTheResponseReallyComeFromPaybox($httpRequest)) {
+            $message = 'Notification invalid: the response doesn\'t come from Paybox.';
+            $this->logger->error($message, $context);
+            throw new HttpResponse($message);
+        }
+
+        return true;
+    }
+
+    /**
+     * @param GetHttpRequest $httpRequest
+     *
+     * @return bool
+     */
+    private function doesTheResponseReallyComeFromPaybox($httpRequest)
+    {
+        // checks paybox signature
+        $this->logger->notice('checkPayboxSignature', [
+            'return' => $this->checkPayboxSignature($httpRequest),
+        ]);
+        if (false === $this->checkPayboxSignature($httpRequest)) {
+            return false;
+        }
+        // checks origin ip address
+        $this->logger->notice('checkOriginIpAddress', [
+            'return' => $this->checkOriginIpAddress($httpRequest),
+        ]);
+        if (false === $this->checkOriginIpAddress($httpRequest)) {
+            return false;
+        }
+
+        return true;
     }
 
+    /**
+     * @return bool
+     */
+    private function checkOriginIpAddress(GetHttpRequest $httpRequest)
```

Comment:
Is this necessary ? The HMAC signature is meant to prevent that kind of checks, what if Paybox add a new IP ? Should the application be impacted ? What about IPv6 ?

I would remove this method to rely only on hmac signature.

Reply:
![Screenshot_20201026_185446](https://user-images.githubusercontent.com/6716686/97209595-d3d25700-17bc-11eb-99b0-d9a5ec393f15.png)

I agree but I followed Paybox documentation

Reply:
Ok, let's keep that for now !